src/coreclr/jit/optimizer.cpp

```diff
@@ -3246,9 +3246,152 @@ bool Compiler::optCanonicalizeExit(FlowGraphNaturalLoop* loop, BasicBlock* exit)
         }
     }
 
+    optSetExitWeight(loop, newExit);
     return true;
 }
 
+//-----------------------------------------------------------------------------
+// optEstimateEdgeLikelihood: Given a block "from" that may transfer control to
+// "to", estimate the likelihood that this will happen taking profile into
+// account if available.
+//
+// Parameters:
+//   from        - From block
+//   to          - To block
+//   fromProfile - [out] Whether or not the estimate is based on profile data
+//
+// Returns:
+//   Estimated likelihood of the edge being taken.
+//
+weight_t Compiler::optEstimateEdgeLikelihood(BasicBlock* from, BasicBlock* to, bool* fromProfile)
```

Comment:
Hopefully this should make the logic more easily updated once we get to the point of having likelihoods here. I also want to switch the preheader code to use this helper in a separate PR.

Reply:
Yes it will become much simpler as you can just query the edge.